joint_space_mdp.py

```diff
@@ -0,0 +1,482 @@
+        self.dx = self.dy = 0.1
+        self.xmin, self.xmax = 0, 0.8
+        self.ymin, self.ymax = 0, 0.8
+        self.X = np.arange(
+            start=self.xmin,
+            stop=self.xmax+self.dx,
+            step=self.dx)
+        self.Y = np.arange(
+            start=self.ymin,
+            stop=self.ymax+self.dy,
+            step=self.dy)
+        self.H, self.W = len(self.Y), len(self.X)
+
+        self.OUT_OF_BOUNDS_COST = MDP.INF
+        self.V = dict()
+        self.P = dict()
+
+    def state_to_key(self, state):
+        # (q1, q2, q3, q4, q5, q6, q7, x, y) = state
+        da = self.A.da_rad
+        joint_state = state[0:self.env.arm.num_joints]
+        joint_indexes = np.rint((joint_state - self.env.arm.ll) / da)
+        # make sure indexes lie within valid bounds
+        joint_indexes = np.clip(joint_indexes, np.zeros_like(
+            joint_indexes), self.max_joint_indexes)
+
+        x, y = state[self.env.arm.num_joints:]
+        xi = int(round((x - self.xmin)/self.dx))
+        yi = int(round((y - self.ymin)/self.dy))
+        # usually leave out-of-bounds as-is, but for special case of rounding
+        # up to bound, just keep within index range
+        if xi == self.W:
+            xi -= 1
+        if yi == self.H:
+            yi -= 1
+
+        key = tuple(np.append(joint_indexes, [xi, yi]))
+        return key
+
+    def solve_mdp(self):
+        np.savez("value_policy_iter_0", V=self.V, P=self.P)
+        for self.main_iter in range(self.max_iters):
+            self.update_policy()
+            self.evaluate_policy()
+            np.savez("results/value_policy_iter_%d" %
+                     (self.main_iter+1), V=self.V, P=self.P)
+            print("Percent complete: %.3f" %
+                  (self.main_iter / float(self.max_iters)))
+
+    def evaluate_policy(self):
+        """Use current policy to estimate new value function
+        """
+        print("Evaluting Policy to Update Value Function...")
+        # new_V = np.ones_like(self.V) * self.OUT_OF_BOUNDS_COST
+        num_states = float(len(self.valid_states))
+        start = time.time()
+        for (x, y) in self.valid_states:
+            self.env.change_bottle_pos(
+                new_pos=[x, y, 0.1],
+                target_type=self.target_type)
+            (xi, yi) = self.state_to_idx((x, y))
+            best_actions = self.P[yi][xi]
+            expected_value = 0
+            # uniform distb for best actions
+            prob = 1./float(len(best_actions))
+            for ai in best_actions:
+                action = self.A[ai]
+                # if desired x,y,z is out of reach, skip this action
+                # validity depends on contact height
+                target_dist = np.linalg.norm(np.array([x, y]))
+                if target_dist > self.env.arm.calc_max_horiz_dist(
+                        action.height):
+                    continue
+
+                value, ns = self.env.run_sim_stochastic(action)
```

Comment:
Where is run_sim_stochastic defined? It looks like it's expected on Environment, but doesn't seem to be there?